File: scripts/lib/impact/keywords.py
"""Keyword extraction and grep-based file matching for impact analysis."""

from __future__ import annotations

import os
import re
import subprocess
from collections import defaultdict

STOPWORDS = frozenset({
    "a", "an", "the", "is", "are", "was", "were", "be", "been", "being",
    "have", "has", "had", "do", "does", "did", "will", "would", "could",
    "should", "may", "might", "shall", "can", "need", "to", "of", "in",
    "for", "on", "with", "at", "by", "from", "as", "into", "through",
    "during", "before", "after", "above", "below", "between", "and", "but",
    "or", "nor", "not", "so", "yet", "both", "either", "neither", "each",
    "every", "all", "any", "few", "more", "most", "other", "some", "such",
    "no", "only", "own", "same", "than", "too", "very", "just", "because",
    "if", "when", "where", "how", "what", "which", "who", "whom", "this",
    "that", "these", "those", "it", "its", "i", "me", "my", "we", "our",
    "you", "your", "he", "him", "his", "she", "her", "they", "them", "their",
    "add", "fix", "update", "remove", "delete", "change", "modify", "make",
    "get", "set", "use", "try", "check", "create", "build", "run",
})


def _split_identifier(token: str) -> list[str]:
    """Split camelCase and snake_case identifiers into parts."""
    # Split on underscores
    parts = token.split("_")
    result = []
    for part in parts:
        # Split camelCase: insert boundary before uppercase letters
        camel = re.sub(r"([a-z0-9])([A-Z])", r"\1_\2", part)
        for sub in camel.split("_"):
            if sub:
                result.append(sub.lower())
    return result


def extract_keywords(prompt: str) -> list[str]:
    """Extract searchable keywords from a natural language prompt.

    Splits camelCase/snake_case identifiers, removes stopwords,
    deduplicates, and lowercases.
    """
    if not prompt.strip():
        return []

    tokens = re.findall(r"[A-Za-z0-9_]+", prompt)
    seen: set[str] = set()
    result: list[str] = []

    for token in tokens:
        parts = _split_identifier(token)
        for part in parts:
            low = part.lower()
            if low not in STOPWORDS and len(low) > 1 and low not in seen:
                seen.add(low)
                result.append(low)

    return result


def grep_matches(root: str, keywords: list[str]) -> dict[str, float]:
    """Find files matching keywords and score by match density.

    Returns dict mapping relative filepath to match score.
    Scores are match count / total lines (density).
    """
    if not keywords or not os.path.isdir(root):
        return {}

    counts: defaultdict[str, int] = defaultdict(int)
    lines: dict[str, int] = {}

    for kw in keywords:
        try:
            result = subprocess.run(
                ["grep", "-r", "-i", "-l",
                 "--exclude-dir=node_modules", "--exclude-dir=.git",
                 "--exclude-dir=.venv", "--exclude-dir=__pycache__",
                 "--exclude-dir=vendor", "--exclude-dir=dist",
                 "--exclude-dir=build", "--exclude-dir=_site",
                 "--include=*.py", "--include=*.js",
                 "--include=*.ts", "--include=*.tsx", "--include=*.jsx",
                 "--include=*.ex", "--include=*.exs",
                 "--include=*.md", "--include=*.json", "--include=*.yaml",
                 "--include=*.yml", "--include=*.toml", "--include=*.cfg",
                 "--include=*.txt", "--include=*.html", "--include=*.css",
                 "--include=*.sh", "--include=*.rs", "--include=*.go",
                 kw, root],
                capture_output=True, text=True, timeout=15,
            )
        except (subprocess.TimeoutExpired, FileNotFoundError):
            continue

        if result.returncode != 0:
            continue

        for path in result.stdout.strip().splitlines():
            path = path.strip()
            if not path:
                continue
            rel = os.path.relpath(path, root)
            counts[rel] += 1

    if not counts:
        return {}

    # Calculate density: count keyword occurrences per file
    scores: dict[str, float] = {}
    for rel, match_count in counts.items():
        full = os.path.join(root, rel)
        if rel not in lines:
            try:
                with open(full, "r", errors="ignore") as f:
                    lines[rel] = max(len(f.readlines()), 1)
            except OSError:
                lines[rel] = 1
        density = match_count / lines[rel]
        scores[rel] = round(density, 6)

    return scores

File: src/impact/keywords.rs
//! Keyword extraction and grep-based file matching.

use regex::Regex;
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use std::process::Command;
use walkdir::WalkDir;

const STOPWORDS: &[&str] = &[
    "a", "an", "the", "is", "are", "was", "were", "be", "been", "to", "of", "in",
    "for", "on", "with", "at", "by", "from", "as", "and", "but", "or", "not", "so",
    "if", "when", "where", "how", "what", "which", "who", "this", "that", "it", "its",
    "add", "fix", "update", "remove", "delete", "change", "modify", "make",
    "get", "set", "use", "try", "check", "create", "build", "run",
];

/// Extract searchable keywords from a natural language prompt.
pub fn extract_keywords(prompt: &str) -> Vec<String> {
    if prompt.trim().is_empty() {
        return Vec::new();
    }

    let re = Regex::new(r"[A-Za-z0-9_]+").unwrap();
    let camel_split = Regex::new(r"([a-z0-9])([A-Z])").unwrap();

    let mut seen = std::collections::HashSet::new();
    let mut result = Vec::new();

    for token in re.find_iter(prompt) {
        let token = token.as_str();
        // Split on underscores
        for part in token.split('_') {
            // Split camelCase
            let expanded = camel_split.replace_all(part, "${1}_${2}");
            for sub in expanded.split('_') {
                let low = sub.to_lowercase();
                if low.len() > 1
                    && !STOPWORDS.contains(&low.as_str())
                    && seen.insert(low.clone())
                {
                    result.push(low);
                }
            }
        }
    }

    result
}

/// Find files matching keywords and score by match density.
pub fn grep_matches(root: &Path, keywords: &[String]) -> HashMap<String, f64> {
    if keywords.is_empty() || !root.is_dir() {
        return HashMap::new();
    }

    let skip = ["node_modules", ".git", ".venv", "__pycache__", "vendor", "dist", "build", "_site", "target"];
    let extensions = [
        "py", "js", "ts", "tsx", "jsx", "ex", "exs", "rs", "go",
        "md", "json", "yaml", "yml", "toml", "sh", "html", "css",
    ];

    let mut counts: HashMap<String, u32> = HashMap::new();

    for kw in keywords {
        let output = Command::new("grep")
            .args(&["-r", "-i", "-l"])
            .args(skip.iter().map(|d| format!("--exclude-dir={d}")))
            .args(extensions.iter().map(|e| format!("--include=*.{e}")))
            .arg(kw)
            .arg(root)
            .output();

        if let Ok(out) = output {
            if out.status.success() {
                for path in String::from_utf8_lossy(&out.stdout).lines() {
                    let trimmed = path.trim();
                    if !trimmed.is_empty() {
                        if let Ok(rel) = pathdiff(trimmed, root) {
                            *counts.entry(rel).or_insert(0) += 1;
                        }
                    }
                }
            }
        }
    }

    if counts.is_empty() {
        return HashMap::new();
    }

    // Calculate density
    let mut scores = HashMap::new();
    for (rel, count) in &counts {
        let full = root.join(rel);
        let lines = fs::read_to_string(&full)
            .map(|c| c.lines().count().max(1))
            .unwrap_or(1);
        scores.insert(rel.clone(), *count as f64 / lines as f64);
    }

    scores
}

fn pathdiff(abs: &str, base: &Path) -> anyhow::Result<String> {
    let abs = Path::new(abs);
    Ok(abs.strip_prefix(base)?.to_string_lossy().to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    #[test]
    fn extract_splits_camel() {
        let kws = extract_keywords("fix AuthService bug");
        assert!(kws.contains(&"auth".to_string()));
        assert!(kws.contains(&"service".to_string()));
    }

    #[test]
    fn extract_splits_snake() {
        let kws = extract_keywords("update user_profile");
        assert!(kws.contains(&"user".to_string()));
        assert!(kws.contains(&"profile".to_string()));
    }

    #[test]
    fn extract_removes_stopwords() {
        let kws = extract_keywords("add the login to the system");
        assert!(!kws.contains(&"the".to_string()));
        assert!(!kws.contains(&"to".to_string()));
    }

    #[test]
    fn extract_empty() {
        assert!(extract_keywords("").is_empty());
    }

    #[test]
    fn grep_finds_files() {
        let tmp = TempDir::new().unwrap();
        let r = tmp.path();
        fs::write(r.join("auth.py"), "class AuthService:\n    pass\n").unwrap();
        fs::write(r.join("db.py"), "class Database:\n    pass\n").unwrap();
        let results = grep_matches(r, &["auth".into()]);
        assert!(results.contains_key("auth.py"));
    }

    #[test]
    fn grep_empty_keywords() {
        let tmp = TempDir::new().unwrap();
        assert!(grep_matches(tmp.path(), &[]).is_empty());
    }
}
